File: src/main/java/phoenix/enity/KnifeEntity.java
package phoenix.enity;

import net.minecraft.entity.EntityType;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.item.ItemEntity;
import net.minecraft.entity.projectile.ThrowableEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.particles.ParticleTypes;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.util.math.EntityRayTraceResult;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.world.World;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import org.jetbrains.annotations.NotNull;
import phoenix.init.PhoenixEntities;
import phoenix.init.PhoenixItems;
import phoenix.items.ash.KnifeItem;

public class KnifeEntity extends ThrowableEntity
{
    public ItemStack knife = new ItemStack(PhoenixItems.ZIRCONIUM_KNIFE.get());
    public boolean isReal = true;
    public static KnifeEntity create(EntityType<? extends ThrowableEntity> type, World worldIn)
    {
        return new KnifeEntity(worldIn, null);
    }

    public KnifeEntity(World worldIn, LivingEntity owner)
    {
        super(PhoenixEntities.KNIFE.get(), owner, worldIn);
    }
    public KnifeEntity(World worldIn, LivingEntity owner, boolean isReal)
    {
        super(PhoenixEntities.KNIFE.get(), owner, worldIn);
        this.isReal = isReal;
    }

    @Override
    @OnlyIn(Dist.CLIENT)
    public void handleStatusUpdate(byte id) {
        if (id == 3) {
            for (int i = 0; i < 8; ++i) {
                //Появляются частицы лавы. В обычном майнкрафте они образуются на потолке, если сверху лава.
                world.addParticle(ParticleTypes.RAIN, getPosX(), getPosY(), getPosZ(), 0.0D, 0.0D, 0.0D);
            }
        }
    }

    @Override
    protected void onImpact(@NotNull RayTraceResult result)
    {
        if(!world.isRemote)
        {
            boolean dropItem = isReal;
            switch (result.getType())
            {
                case ENTITY:
                    dropItem &= ((KnifeItem) knife.getItem()).onHitEntity(world, owner, this, ((EntityRayTraceResult)result).getEntity(), knife);
                    break;
                case BLOCK:
                    dropItem &= ((KnifeItem) knife.getItem()).onHitBlock(world, owner, ((BlockRayTraceResult)result).getPos(), this, knife);
                    break;
            }
            if (dropItem)  world.addEntity(new ItemEntity(world, getPosX(), getPosY(), getPosZ(), knife));

            onKillCommand();
        }
    }

    @Override
    protected void registerData()
    {

    }
}

File: src/main/java/phoenix/enity/KnifeEntity.kt
package phoenix.enity

import net.minecraft.block.Blocks
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.enchantment.Enchantments
import net.minecraft.entity.EntityType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.item.ItemEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.projectile.ThrowableEntity
import net.minecraft.item.ItemStack
import net.minecraft.particles.ParticleTypes
import net.minecraft.util.math.BlockRayTraceResult
import net.minecraft.util.math.EntityRayTraceResult
import net.minecraft.util.math.RayTraceResult
import net.minecraft.util.math.RayTraceResult.Type.BLOCK
import net.minecraft.util.math.RayTraceResult.Type.ENTITY
import net.minecraft.world.World
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.fml.network.NetworkHooks
import phoenix.init.PhoenixEnchantments
import phoenix.init.PhoenixEntities
import phoenix.init.PhoenixItems
import phoenix.items.ash.KnifeItem
import phoenix.utils.getEnchantmentLevel


class KnifeEntity : ThrowableEntity
{
    var knife = ItemStack(PhoenixItems.ZIRCONIUM_KNIFE.get())
    var isReal = true
    constructor(type: EntityType<KnifeEntity>, worldIn: World) : super(type, worldIn)
    constructor(worldIn: World, owner: LivingEntity, isReal: Boolean) : super(PhoenixEntities.KNIFE.get(), owner, worldIn)
    {
        this.isReal = isReal;
    }

    override fun tick()
    {
        super.tick()
        if(EnchantmentHelper.getEnchantmentLevel(Enchantments.FLAME, knife) > 0)
        {
            world.addParticle(ParticleTypes.LAVA, posX, posY, posZ, 0.0, 0.0, 0.0)
        }
    }

    @OnlyIn(Dist.CLIENT)
    override fun handleStatusUpdate(id: Byte)
    {
        if (id.toInt() == 3)
        {
            for (i in 0..7)
            {
                //Появляются частицы лавы. В обычном майнкрафте они образуются на потолке, если сверху лава.
                world.addParticle(ParticleTypes.RAIN, posX, posY, posZ, 0.0, 0.0, 0.0)
            }
        }
    }

    override fun onImpact(result: RayTraceResult)
    {
        if (!world.isRemote)
        {
            var dropItem = isReal
            when (result.type)
            {
               ENTITY ->
               {
                   dropItem = dropItem && (knife.item as KnifeItem).onHitEntity(world, owner, this, (result as EntityRayTraceResult).entity, knife)
                   knife.attemptDamageItem(1, rand, null)
                   if (knife.getEnchantmentLevel(Enchantments.LOYALTY) > 0 && owner is PlayerEntity && isReal)
                       (owner as PlayerEntity).addItemStackToInventory(knife)
                   else if (dropItem) world.addEntity(ItemEntity(world, posX, posY, posZ, knife))

                   onKillCommand()
               }
               BLOCK ->
               {
                   val block = world.getBlockState(((result as BlockRayTraceResult).pos))
                   dropItem = dropItem and (knife.item as KnifeItem).onHitBlock(world, owner, result.pos, this, knife)
                   if(block.block !== Blocks.GRASS || block.block !== Blocks.TALL_GRASS)
                   {
                       knife.attemptDamageItem(1, rand, null)
                       if (knife.getEnchantmentLevel(Enchantments.LOYALTY) > 0 && owner is PlayerEntity && isReal)
                           (owner as PlayerEntity).addItemStackToInventory(knife)
                       else if (knife.getEnchantmentLevel(PhoenixEnchantments.TELEPORTATION.get()) > 0)
                           (owner as PlayerEntity).addItemStackToInventory(knife)
                       else if (dropItem)
                           world.addEntity(ItemEntity(world, posX, posY, posZ, knife))
                       onKillCommand()
                   }
               }
                else -> {}
            }
        }
    }

    override fun registerData()
    {
    }

    override fun createSpawnPacket() = NetworkHooks.getEntitySpawningPacket(this)!!
}
